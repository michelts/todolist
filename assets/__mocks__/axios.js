export default {
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
};
