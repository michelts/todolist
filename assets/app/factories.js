import Factory from 'rosie';

const UserFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('username', (index) => `username_${index}`)
  .sequence('name', (index) => `Name-${index}`);

export {
  UserFactory,
};
