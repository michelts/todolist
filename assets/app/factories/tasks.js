import { Factory } from 'rosie';

const TaskFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('description', (index) => `Task ${index}`)
  .attr('priority', 1)
  .attr('due_date', '2020-05-22')
  .attr('is_completed', false);

export {
  TaskFactory,
}
