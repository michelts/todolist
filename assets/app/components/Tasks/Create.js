import React from 'react';
import axios from 'axios';
import Button from 'react-bootstrap/Button';

const CreateButton = ({ tasks, setTasks }) => {
  const [creating, setCreating] = React.useState(false);

  const handleClick = () => {
    setCreating(true);
    axios.post('/api/v1/tasks/', { description: '' }).then(({ data }) => {
      setTasks((state) => state.set(data.id, data));
      setCreating(false);
    });
  };

  return (
    <Button
      disabled={tasks === undefined || creating}
      onClick={handleClick}
    >
      Add new task
    </Button>
  );
};

export default CreateButton;
