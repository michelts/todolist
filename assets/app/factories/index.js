import { UserFactory } from './user';
import { TaskFactory, BlankTaskFactory } from './tasks';

export {
  UserFactory,
  TaskFactory,
  BlankTaskFactory,
};
