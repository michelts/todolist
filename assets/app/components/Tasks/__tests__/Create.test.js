import React from 'react';
import axios from 'axios';
import { OrderedMap } from 'immutable';
import { shallow } from 'enzyme';
import { TaskFactory } from '../../../factories';
import Create from '../Create';

const getWrapper = (customProps = {}) => {
  const props = {
    disabled: false,
    setTasks: jest.fn(),
    ...customProps,
  };
  const wrapper = shallow(<Create {...props} />);
  return { wrapper, props };
};

describe('Create component', () => {
  it('should render a disabled button if disabled flag is provided', () => {
    const { wrapper } = getWrapper({ disabled: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a button to add new task if disabled is not true', () => {
    const { wrapper } = getWrapper({ disabled: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('should disable the create button, create a blank task and append it to the end of the list', async () => {
    const newTask = TaskFactory.build({ description: '' });
    axios.post.mockResolvedValue({ data: newTask });

    const tasks = OrderedMap();
    const { wrapper, props: { setTasks } } = getWrapper({ tasks });

    wrapper.simulate('click');
    expect(wrapper.prop('disabled')).toBe(true);
    expect(axios.post).toHaveBeenCalledWith('/api/v1/tasks/', { description: '' });

    await axios.post;
    const callback = setTasks.mock.calls[0][0];
    const newTasks = callback(OrderedMap()); // simulate callback over a clean state
    expect(newTasks).toEqual(OrderedMap([[newTask.id, newTask]]));
    expect(wrapper.prop('disabled')).toBe(false);
  });
});
