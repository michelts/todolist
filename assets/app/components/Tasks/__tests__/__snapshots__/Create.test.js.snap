// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create component should render a button to add new task if disabled is not true 1`] = `
<Button
  active={false}
  disabled={false}
  onClick={[Function]}
  type="button"
  variant="primary"
>
  Add new task
</Button>
`;

exports[`Create component should render a disabled button if disabled flag is provided 1`] = `
<Button
  active={false}
  disabled={true}
  onClick={[Function]}
  type="button"
  variant="primary"
>
  Add new task
</Button>
`;
